src/dist/mod.rs

```diff
@@ -877,45 +877,68 @@ impl fmt::Display for Profile {
 }
 
 pub(crate) struct DistOptions<'a> {
-    pub(crate) cfg: &'a Cfg<'a>,
-    pub(crate) toolchain: &'a ToolchainDesc,
-    pub(crate) profile: Profile,
-    pub(crate) update_hash: Option<&'a Path>,
-    pub(crate) dl_cfg: DownloadCfg<'a>,
+    pub(super) cfg: &'a Cfg<'a>,
+    pub(super) toolchain: &'a ToolchainDesc,
+    profile: Profile,
+    pub(super) update_hash: PathBuf,
+    dl_cfg: DownloadCfg<'a>,
     /// --force bool is whether to force an update/install
-    pub(crate) force: bool,
+    force: bool,
     /// --allow-downgrade
-    pub(crate) allow_downgrade: bool,
+    pub(super) allow_downgrade: bool,
     /// toolchain already exists
-    pub(crate) exists: bool,
+    pub(super) exists: bool,
     /// currently installed date and version
-    pub(crate) old_date_version: Option<(String, String)>,
+    pub(super) old_date_version: Option<(String, String)>,
     /// Extra components to install from dist
-    pub(crate) components: &'a [&'a str],
+    components: &'a [&'a str],
     /// Extra targets to install from dist
-    pub(crate) targets: &'a [&'a str],
+    targets: &'a [&'a str],
+}
+
+impl<'a> DistOptions<'a> {
+    pub(super) fn new(
+        components: &'a [&'a str],
+        targets: &'a [&'a str],
+        toolchain: &'a ToolchainDesc,
+        profile: Profile,
+        force: bool,
+        cfg: &'a Cfg<'_>,
+    ) -> Result<Self> {
```

Comment:
Hm, I'm slightly nervous about the number of arguments here, especially as the first two have the same type and `force` is just a `bool` (making it not immediately obvious what it means without looking it up). Does the new function pull its weight? Looking at the current uses it seems fine as they're mostly passing arguments on but I did want to query this.

If we had some kind of partial `Default` I'd suggest using that but we don't.